[MM-2541] Shortcut to mark all channels as read for a team

joshuadschoep opened the pull request:
#### Summary

This effort adds the shortcut `Shift` + `ESC` from  logged-in pages w/ a sidebar that marks all channels in the sidebar as read. This will, for now, only work on desktop.

It also adds a small toast to explain the explanation upon enabling via feature flag (`MM_FEATUREFLAGS_ENABLESHIFTESCAPETOMARKALLREAD`).

Most of the information for this ticket can be found in Jira:
https://mattermost.atlassian.net/browse/MM-2541
Along with Figma designs for the new features.

##### Testing Notes

3 changes have been made with this effort:

1. Once the feature is enabled, a toast explaining the new shortcut key will appear at the top right of the browser/app (Screenshot below).
2. A short description has been added to the Help -> Keyboard Shortcuts menu under the Messages header.
3. If enabled, if a user pressed Shift + ESC on any page with a Sidebar, a modal will appear asking them to confirm whether to mark all chats, channels, messages as read (Screenshot below). This modal will contain a "Don't ask me again" checkbox.

If the "Got it" or `x` buttons on the toast are clicked, the toast should not appear again on this device unless the user clears their backend-stored preferences.

If the "Don't show me again" checkbox is checked when the user confirms to mark all messages as read, then subsequent `Shift` + `ESC` calls will automatically clear all messages. This persists on the device unless the user clears their backend-stored preferences.
_Note: to keep Cancel 'pure' and without side-effect, no behavior is changed even if the user checks "Don't show me again" and then cancels._

If the user confirms to clear all messages, an API call is sent _only if they have unread messages_, and all messages for the active team are marked as read.

#### Ticket Link
https://github.com/mattermost/mattermost/issues/20952
https://mattermost.atlassian.net/browse/MM-2541

#### Screenshots
| Label | Close-up | Page |
| --- | --- | --- |
| Introductory Toast | <img width="398" height="209" alt="2025-09-30-222834_hyprshot" src="https://github.com/user-attachments/assets/e6a18ff8-24d5-4499-9aa4-d256971afa1e" /> | <img width="1249" height="360" alt="2025-09-30-222810_hyprshot" src="https://github.com/user-attachments/assets/302f4ff1-4a4b-4ed8-92aa-9c0ca518c245" />|
| Shortcut List | <img width="356" height="220" alt="2025-09-30-223444_hyprshot" src="https://github.com/user-attachments/assets/d5cda640-2d78-4671-9ba2-27475c72791e" /> | <img width="1106" height="779" alt="2025-09-30-223448_hyprshot" src="https://github.com/user-attachments/assets/b21a5bec-0ef8-41b9-89d8-d41f268b5cf8" /> |
| Confirmation Modal | <img width="534" height="312" alt="2025-09-30-222905_hyprshot" src="https://github.com/user-attachments/assets/e334d4e3-674f-406c-9bed-d4066fdf92b6" /> | <img width="1250" height="876" alt="2025-09-30-222853_hyprshot" src="https://github.com/user-attachments/assets/b1916b39-1e20-4ef8-8093-b0524b380192" /> |

These can be cross referenced with [Figma](https://www.figma.com/design/pnWkf5SAoWcHGxMJvVddIP/MM-2541-Mark-All-Channels-Read?node-id=525-22685&t=I6GQEz3acVznpjWa-0) and should match up with two grammar exceptions:
- Added a comma after "Don't worry" in the introductory toast.
- Added commas (including oxford) in between "channels, threads, and Direct Messages" in the confirmation modal.

#### Release Note
```release-note
Added a new shortcut, `Shift` + `ESC`, that will mark all channels, threads, and direct messages as read for a team.
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Change Impact: 🟡 Medium

**Regression Risk:** Changes touch server APIs, DB store queries, app logic, Redux actions, and multiple React components (including unread/mention propagation and CRT thread handling), which creates moderate risk of subtle regressions in read-state propagation across UIs; feature-flag gating and added tests mitigate but don't eliminate the risk.  
**QA Recommendation:** Perform targeted manual QA on CRT vs non-CRT thread read behavior, LHS channel/Threads/DM badge consistency after using the shortcut, permission edge cases (marking on behalf of other users), persistence of toast and “Don’t ask me again” preference, and feature-flag toggling; rely on automated tests for baseline coverage but prioritize exploratory checks around thread visibility and unread propagation.

*Generated by CodeRabbitAI*
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

joshuadschoep committed:
feat(webapp): added keyboard shortcut for Mark All As Read (MM-2541)

- Added shortcut (within sidebar) for Shift+ESC to mark _all_ messages, teams as read
    - Desktop only
- Added feature toasts for new features and localStorage support
- Added feature toast for mark-all-as-read feature
    - Should decide when/how people want this shown, I just followed designs
    - Will only show if the user has not clicked 'Got it' before, and is not on mobile
- Added confirmation modal for mark all as read shortcut
    - Contains option to not show again, saved in localStorage
- Added English translations for read shortcut
    - Will need i18n aid on other languages

This is a draft version of this feature update that still needs testing and i18n support, along with a11y validation.

joshuadschoep committed:
feat(webapp): feature flags and fixes for mark all as read shortcut

- Added feature flags surrounding rollout of mark-all-as-read shortcut
- Added shortcut to list of shortcuts in help section
- Extended tests for new components
- Updated snapshot for sidebar_list, keyboard_shortcuts_modal
- Fixed styling and CSS issues

Still in draft, needs documentation and e2e support.

joshuadschoep committed:
Merge branch 'master' into feature/MM-2541

jgheithcock commented on webapp/channels/src/stores/local_storage_store.ts:
```diff
@@ -167,6 +169,27 @@ class LocalStorageStoreClass {
+     * "Do not show again" checkbox. This is used to ensure that
+     * functionality between sessions.
+     */
+    setMarkAllReadWithoutConfirm(confirm: boolean) {
```
These will all need the userID of the current user, otherwise the first user sets it for everyone.

joshuadschoep replied:
Thanks for noticing this, updated within the change to `useGlobalState` as well and tested again, should be good now!

jgheithcock reviewed (changes requested):
Looks great overall, one issue and two observations:
- The code doesn't incorporate the current user id so the first person that comes through sets things for everyone.
- `LocalStorageStore` is a much-used, but deprecated, pattern, `useGlobalState` would save you a lot of code and is a better, more 'Redux' pattern.
- This introduces yet-another Toast-style component, we already have Toast and InfoToast. It might have been less work for you (and a better design) to wrap InfoToast.

That said, this was really two requests in one, we should start using this `FeaturesToAnnounce` and `FeatureToast` moving forward. Nice job making this expandable.

joshuadschoep committed:
fix(webapp): fixed some issues with new mark-all-read feature

- Scoped persistent storage to current user ID
  so that subsequent new logins also get the notification
- Replaced LocalStorage calls with useGlobalState calls, sad
  that I missed that this updated call was being used.
- Fixed an issue that would have caused the new shortcut to
  show up in the Help menu's shortcuts without being enabled.

joshuadschoep committed:
Merge branch 'master' into feature/MM-2541

joshuadschoep commented on webapp/channels/src/components/sidebar/sidebar_list/sidebar_list.tsx:
```diff
@@ -338,6 +351,19 @@ export class SidebarList extends React.PureComponent<Props, State> {
+    markAllChannelsAsReadShortcut = (e: KeyboardEvent) => {
+        if (!e.altKey && e.shiftKey && !e.ctrlKey && !e.metaKey && isKeyPressed(e, Constants.KeyCodes.ESCAPE)) {
+            e.preventDefault();
+            if (LocalStorageStore.getMarkAllReadWithoutConfirm()) {
```
Sad I missed this, but thanks for pointing me in the right direction. Both components should be updated now.

joshuadschoep committed:
Fixed a snapshot test and a missing i18n member

jgheithcock reviewed (changes requested):
My mistake, I should have said to use the `usePreference` hook, to save in the persistent user preferences on the server side.

joshuadschoep committed:
Merge branch 'master' into feature/MM-2541

joshuadschoep committed:
Replaced useGlobalState with backend-ready usePreference. Previous version was just a mistake as we didnt know about the supported API

jgheithcock reviewed (approved):
This all worked for me. Specifically, created two users. Got both "Got it!" and "Mark all messages as read?" for both users. Verified logging in on different browser works as expected.

joshuadschoep committed:
fix(server): fix lint issue with gofmt

lindy65 commented:
Hi @joshuadschoep 

While testing, I came across a few issues:

1. From `user-1` If I **reply** to a message from `sysadmin`


<img width="959" height="467" alt="image" src="https://github.com/user-attachments/assets/56f61227-5c26-49c0-877b-581a2b409f95" />



`sysadmin's` Town Square channel does not bold to show an unread. However, the new message does show up in the "Threads" view. If I send **a new message** in a channel, the channel does bold in the left-hand-side (LHS)
    
    
<img width="809" height="392" alt="image" src="https://github.com/user-attachments/assets/74de3432-2d73-43b4-90c3-038a6d513ea6" />


2. If I press `SHIFT` + `Esc` and mark all threads, messages, DMs as read, only channels and DMs are marked as read. Threads remain showing unread messages. Below are 2 screenshots: 1st is `sysadmin's` screen before pressing `SHIFT` + `Esc` and 2nd is after pressing `SHIFT` + `Esc` - showing that Threads are still unread. Only channels and DMs in the LHS are marked as read. If I press `SHIFT` + `Esc` from the Threads view, the unreads in Threads are still not marked as read.


<img width="507" height="461" alt="image" src="https://github.com/user-attachments/assets/cecd7680-0cd3-4917-8509-8a10c346bb54" />


<img width="539" height="407" alt="image" src="https://github.com/user-attachments/assets/2df23ab4-1950-449a-a851-372bab8c872e" />

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

joshuadschoep committed:
feat(server,webapp): added cleaner and more effective method with which to mark-all-read

- Added 2 new routes to the API (need to find docs to update those):
    - `PUT /api/v4/channels/members/<userId>/direct/read` will mark a user's non-team DMs and GMs as read
    - `PUT /api/v4/users/<userId>/teams/<teamId>/read` will do a similar action as the multi-channel mark_read action, but with a teamId signifier. Because this is using a teamId, it will _not_ handle DMs or GMs.
- Updated sidebar_list.tsx to use these new routes for the new shortcut
- Added extensive testing, including feature flag assurance.

joshuadschoep commented:
Hi all!

Got some time to set aside for this and realized with needing to handle threads, that the FE logic would just be a lot cleaner if we had a couple dedicated API hits for this project. To be clear, the previous architecture did indeed update threads, but never went back and reported those as updated to the FE. Some weirdness with `readMultipleChannels` that might be worth looking into. But, at the time, I definitely did not have CRT enabled nor did I know of its existence, hence not knowing to test it, apologies!

These updates do much as the commit message says:

- Added 2 new routes to the API (need to find docs to update those):
    - `PUT /api/v4/channels/members/<userId>/direct/read` will mark a user's non-team DMs and GMs as read
    - `PUT /api/v4/users/<userId>/teams/<teamId>/read` will do a similar action as the multi-channel mark_read action, but with a teamId signifier. Because this is using a teamId, it will _not_ handle DMs or GMs.
- Updated sidebar_list.tsx to use these new routes for the new shortcut

Tested many times over myself and it should be marking all of threads, DMs, GMs, and normal channels as read. Threads and channels will be _on a team-wide basis_, while DMs and GMs will not (they are not part of any team within the models).

mattermost-build committed:
Merge branch 'master' into feature/MM-2541

mattermost-build committed:
Merge branch 'master' into feature/MM-2541

lieut-data commented:
/update-branch

mattermost-build committed:
Merge branch 'master' into feature/MM-2541

lieut-data committed:
fix from upstream changes

lieut-data commented:
Thanks, @lindy65! I've pushed again to align with upstream changes.

lindy65 commented:
Thanks @lieut-data - I still see CI failing checks. Is there anything that can be done to get this PR across the line?

lieut-data committed:
fix: eslint errors in teams actions

lieut-data committed:
document new API endpoints

lieut-data committed:
fix i18n

lindy65 reviewed (approved):
Thanks @joshuadschoep - looks good to me 👍 I spun up my own `/cloud` server as I couldn't get a test server up on the PR. If anyone wants the credentials for my test server, just shout.

@calebroseland - just you to review now thanks :)

calebroseland commented on server/channels/app/channel.go:
```diff
@@ -3162,7 +3162,113 @@ func (a *App) SearchChannelsUserNotIn(rctx request.CTX, teamID string, userID st
+		case errors.As(err, &invErr):
+			return nil, model.NewAppError("MarkTeamChannelsAndThreadsViewed", "app.channel.update_last_viewed_at.app_error", nil, "", http.StatusBadRequest).Wrap(err)
+		default:
+			return nil, model.NewAppError("MarkTeamChannelsAndThreadsViewed", "app.channel.update_last_viewed_at.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
```
```suggestion
			return nil, model.NewAppError("MarkAllDirectAndGroupMessagesViewed", "app.channel.update_last_viewed_at.app_error", nil, "", http.StatusBadRequest).Wrap(err)
		default:
			return nil, model.NewAppError("MarkAllDirectAndGroupMessagesViewed", "app.channel.update_last_viewed_at.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
```

calebroseland committed:
fix err id

calebroseland committed:
remove unused localhost methods

calebroseland committed:
use ShortcutKey and ShortcutSequence

calebroseland committed:
feature_enhancements, mark as read toast enchancements

calebroseland committed:
read all modal mount point, use openModal

calebroseland committed:
use handler

calebroseland committed:
fix style

calebroseland reviewed (approved):
Thanks @joshuadschoep! This was already looking great, but I noticed a couple of smaller items that could use some polish. Apologies if I went too far with changes, but I ended up just committing the explorations I had done. 

There were mainly only two changes I was going to request:
- 2f0aec8e207a9d0820ddddf941229607b2bac828 fixes a mis-matched AppError ID
- db43e838ed187f3c5ccc0b6c565a133ea7a61cdc removes unused localstorage methods

And then a few non-blockers/suggestions/polish I ended up just committing after some review/exploration:
- d8600a6c12562d87123119f3056d6cf1072874f1: reworks existing `ShortcutKey` mapping into a new `ShortcutSequence` component which handles mapping internally; this and some existing msgs are used to simplify `{shift} {escape}` into `{shortcut}` for i18n.
- 312a14114f744e4f70a5e06ed2bfcb6362f4f2cb and  c1be96e3f5252c13d681e40b37cab724ed5eb122: 
  - `MarkAllAsReadToast` moved to `ChannelController`, which keeps the toast in channels and prevents it from appearing in system console and other places. It also get's portaling, uses a predefined z-index, close button tooltip and hover styles, and hides when other modals are open (e.g. keyboard shortcuts).
  - `MarkAllAsReadModal` uses `openModal`/`ModalController` infra instead of manual mounting/visibility in `sidebar`

Again—this was already in a pretty good state, just adding a bit in the spirit of collaboration!

agarciamontoro commented on server/channels/app/channel.go:
```diff
@@ -3162,7 +3162,113 @@ func (a *App) SearchChannelsUserNotIn(rctx request.CTX, teamID string, userID st
+	// push notifications, but threads do not. As a result, this method call acts on _all threads_, not just out of date ones.
+	// This is, however, a bit of an assumption on my part.
+	// See `MarkChannelsAsViewed` vs. `thread_store.go:MarkAllAsReadByTeam` for what I mean.
+	err = a.Srv().Store().Thread().MarkAllAsReadByTeam(userID, teamID)
```
It does seem risky, indeed. I've been looking at the Thread store, and maybe we can use `SqlThreadStore.MarkAllAsReadByChannels` on the `channelsToView` returned by the call to `GetTeamChannelsWithUnreadAndMentions`? Not really sure if this fits our need here, I have not tested it, but if it does, at least we're isolating to unread channels, which seem slightly less risky.

Another alternative, probably better, is to use `SqlThreadStore.GetTeamsUnreadForUser`, and mark only those as read. That's an additional select query, but I have a feeling that would be better than engineer a single query to get everything we need.

agarciamontoro replied:
To wrap this up: addressed in [`f85332b471`](https://github.com/mattermost/mattermost/commit/f85332b471). We no longer call the unbounded `MarkAllAsReadByTeam` and instead pass channel IDs to `MarkAllAsReadByChannels`, whose `WHERE Threads.LastReplyAt > ThreadMemberships.LastViewed` clause keeps the actual UPDATE bounded to genuinely stale rows even when we hand it every team channel.

agarciamontoro reviewed (changes requested):
Incredible work, @joshuadschoep! I requested changes only to discuss the concern on `MarkAllAsReadByTeam`. Thank you!

calebroseland committed:
fix: fix refactoring typo

agarciamontoro committed:
Merge branch 'master' into feature/MM-2541

agarciamontoro committed:
Merge fix: realign branch with upstream changes

Upstream MM-67319/MM-67320 (#36037) moved ShortcutKey and
WithTooltip into the shared package and rewrote the keyboard
shortcuts test to snapshot real DOM instead of a
react-test-renderer tree. The merge resolution missed several
follow-on consequences; clean them up so the branch builds, type
checks, lints, passes i18n-extract-check and runs without
throwing at mount.

- Port the inline-content variant from the deleted channels-side
  shortcut_key.scss to the new shared shortcut_key.css.
- Refresh the keyboard_shortcuts_sequence snapshot so it matches
  Testing Library's container output (DOM only, no component
  nodes, class= not className=).
- Repoint mark_all_as_read_modal and mark_all_as_read_toast at
  components/shortcut_key for ShortcutKeys and use
  ShortcutKeys.escape; the channels-side with_tooltip is now a
  thin re-export and the field was renamed in the shared keys
  map. Without this both consumers threw "Cannot read properties
  of undefined" at mount.
- Switch mark_all_as_read_toast's UserAgent import to
  @mattermost/shared/utils/user_agent; the channels-local
  utils/user_agent path no longer resolves.
- Drop the orphan mark_all_threads_as_read_modal.cancel string
  from en.json so formatjs extraction is in sync.

agarciamontoro committed:
Clean up TestReadAllInTeam

Drop four lines left from debugging and replace them with a real
assertion: LastViewedAtTimes must contain the test channel with a
value at or after the most recent post.

Update three client.GetChannel calls to the (ctx, id) signature;
the prior etag argument no longer compiles after upstream removed
it.

agarciamontoro committed:
Use SelectBuilder for team channels query

GetTeamChannelsWithUnreadAndMentions built a squirrel query and
then manually called ToSql before handing the string+args to
GetReplica().Select. SelectBuilder accepts the builder directly
and removes the intermediate dance, matching the pattern used
elsewhere in this store.

agarciamontoro committed:
Mark all team-channel threads on team read

MarkTeamChannelsAndThreadsViewed used Thread().MarkAllAsReadByTeam
unconditionally, writing every thread membership in the team for
the user even when nothing was stale. Scoping the call to
channelsToView (channels with unread channel-level messages) would
have closed the perf concern but introduced a regression: in CRT
mode a thread reply does not bump the channel's TotalMsgCount, so
a channel can be read at the channel level while still having
unread thread replies, and those would have been silently skipped.

Build the channel-id list from the keys of the times map instead.
GetTeamChannelsWithUnreadAndMentions already populates that map
for every team channel the user belongs to, so no extra query is
needed. MarkAllAsReadByChannels then filters the actual UPDATE
through its LastReplyAt > LastViewed clause, keeping writes
bounded to genuinely stale rows.

Gate the channel-level work (UpdateLastViewedAt, push clearing,
the MultipleChannelsViewed event) on channelsToView being
non-empty, but always run the thread mark and broadcast
ThreadReadChanged for every team channel so CRT clients refresh
thread state in channels that had no channel-level change.

mattermost-build committed:
Merge branch 'master' into agarciamontoro/mm-2541-merge-validation

agarciamontoro reviewed (commented):
I've addressed the remaining issue with `MarkAllAsReadByTeam` (and the minor nit with the `SelectBuilder` as well).

@jgheithcock, @lieut-data, please take a look when you can :)

agarciamontoro commented:
I'm also seeing that now that coderabbit is automatically reviewing PRs, it found a couple of things. I'll try to address those as well.

jgheithcock reviewed (approved):
Other than what has already been called out, looks great to me. Thanks for taking this on Alejandro!

lieut-data commented on server/channels/app/channel.go:
```diff
@@ -3304,7 +3304,117 @@ func (a *App) SearchChannelsUserNotIn(rctx request.CTX, teamID string, userID st
+
+		for _, channelID := range channelsToClearPushNotifications {
+			a.clearPushNotification(currentSessionID, userID, channelID, "")
+		}
```
We only iterate this if `if len(channelsToView) > 0`. While this is possibly always the case, I think we should just iterate this unconditionally (outside the if block).

agarciamontoro replied:
Fixed in [`06abe0a4c7`](https://github.com/mattermost/mattermost/commit/06abe0a4c7).

lieut-data commented on server/channels/app/channel.go:
```diff
@@ -3304,7 +3304,117 @@ func (a *App) SearchChannelsUserNotIn(rctx request.CTX, teamID string, userID st
+
+	if isCRTEnabled {
+		// Thread updates can have happened in any team channel (not just channelsToView),
+		// so notify clients to refresh thread state across the entire team.
```
Does this trigger the client to /make/ a request back to the server? (The comment implies "refreshing" which implies "API call" which would be problematic if we spawn N client requests.)

agarciamontoro replied:
No, the handler dispatches Redux state updates only. Fixed comment in [`06abe0a4c7`](https://github.com/mattermost/mattermost/commit/06abe0a4c7).

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+		"Channels.TeamId":       teamID,
+		"ChannelMembers.UserId": userID,
+	})
+
```
nit: can we format this differently to clarify the structure??

```go
	query := s.getQueryBuilder().
        Select(
    		"Channels.Id",
	    	"Channels.Type",
	    	"Channels.TotalMsgCount",
	    	"Channels.LastPostAt",
	    	"ChannelMembers.MsgCount",
	    	"ChannelMembers.MentionCount",
	    	"ChannelMembers.NotifyProps",
	    	"ChannelMembers.LastViewedAt",
	    ).From("ChannelMembers").
		InnerJoin("Channels ON ChannelMembers.ChannelId = Channels.Id").
        Where(sq.Eq{
		    "Channels.TeamId":       teamID,
		    "ChannelMembers.UserId": userID,
	    })
```

agarciamontoro replied:
Fixed in [`f9924bda8a`](https://github.com/mattermost/mattermost/commit/f9924bda8a).

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+
+	channelsWithUnreads := []string{}
+	channelsWithMentions := []string{}
+	readTimes := map[string]int64{}
```
Should we pre-allocate this as `len(channels)`?

agarciamontoro replied:
Fixed in [`f9924bda8a`](https://github.com/mattermost/mattermost/commit/f9924bda8a).

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+	readTimes := map[string]int64{}
+
+	for i := range channels {
+		channel := channels[i]
```
Should we `for _, channel := range channels` instead?

agarciamontoro replied:
Fixed in [`f9924bda8a`](https://github.com/mattermost/mattermost/commit/f9924bda8a).

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+		if notify == model.ChannelNotifyDefault {
+			notify = userNotifyProps[model.PushNotifyProp]
+		}
+		if notify == model.UserNotifyAll || channel.Type == string(model.ChannelTypeDirect) {
```
I haven't exhaustively reviewed the logic here, but it strikes me as odd that we haven't already pulled this out into a helper function or something that's not buried deep with a specific query.

agarciamontoro replied:
Filed [MM-68703](https://mattermost.atlassian.net/browse/MM-68703) to extract the shared post-query helper across `GetTeamChannelsWithUnreadAndMentions`, `GetDirectMessagesWithUnreadAndMentions`, and the pre-existing `GetChannelsWithUnreadsAndWithMentions`. Will follow up there rather than land it on top of this feature branch.

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+	return channelsWithUnreads, channelsWithMentions, readTimes, nil
+}
+
+func (s SqlChannelStore) GetMessagesWithUnreadAndMentions(rctx request.CTX, userID string, userNotifyProps model.StringMap) ([]string, []string, map[string]int64, error) {
```
`GetMessagesWithUnreadAndMentions` is ambiguous -- should we call this `GetDirectMessagesWithUnreadAndMentions`?

agarciamontoro replied:
Fixed in [`f9924bda8a`](https://github.com/mattermost/mattermost/commit/f9924bda8a).

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+		"ChannelMembers.NotifyProps",
+		"ChannelMembers.LastViewedAt",
+	).From("ChannelMembers").
+		InnerJoin("Channels ON ChannelMembers.ChannelId = Channels.Id").Where(sq.Eq{
```
FWIW, I usually just write `Join` and not `InnerJoin`. Functionally the same, but I always have to pause and ask what's "special" about the inner.

agarciamontoro replied:
Fixed in [`f9924bda8a`](https://github.com/mattermost/mattermost/commit/f9924bda8a).

lieut-data commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+			},
+		})
+
+	queryString, args, err := query.ToSql()
```
Can we use `SelectBuilder` instead?

agarciamontoro replied:
Fixed in [`f9924bda8a`](https://github.com/mattermost/mattermost/commit/f9924bda8a).

lieut-data reviewed (changes requested):
I focused on the server-side changes only -- just a handful of comments below.

agarciamontoro committed:
Mark mark-read audit records as success

The handlers for mark all DM/GM and mark team read created an
audit record with status Fail and never updated it on success,
so successful calls were always logged as failures.

agarciamontoro committed:
Mark all DM/GM threads on full read

MarkAllDirectAndGroupMessagesViewed early-returned when no
channel had unreads, so followed threads in DMs/GMs whose
channel-level counters were already current stayed unread under
CRT. Mirror MarkTeamChannelsAndThreadsViewed and call
MarkAllAsReadByChannels for every DM/GM in times.

agarciamontoro committed:
Polish DM/GM channels-with-unreads query

Use model.ChannelTypeDirect/Group constants instead of bare
"D"/"G" literals, and update the error wrap to mention DM/GM
channels (it was copied from the team variant).

agarciamontoro committed:
Fix stale ReadAllMessages godoc

agarciamontoro committed:
Type last_viewed_at_times as int64 map in OpenAPI

The response field was declared as a generic object. Add
additionalProperties so generated clients see it as a
channelId -> int64 timestamp map.

agarciamontoro committed:
Gate MarkAllAsReadToast mount on feature flag

The toast was mounted unconditionally, so its async chunk loaded
even when EnableShiftEscapeToMarkAllRead was off. Gate the mount
with the flag so the chunk only loads when the feature is on.

agarciamontoro committed:
Return data from markAllInTeamAsRead thunk

Match the {data: response} shape used by adjacent thunks instead
of returning {}, so callers can read the API payload.

agarciamontoro committed:
Coerce undefined suffix in createStoredKey

createStoredKey('foo') returned 'fooundefined' when the suffix
arg was omitted. Coerce a missing suffix to ''.

agarciamontoro committed:
Refactor mark-read websocket events

agarciamontoro committed:
Polish DM/GM channels-with-unreads query

agarciamontoro committed:
Merge branch 'master' into feature/MM-2541

agarciamontoro commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2132,6 +2132,153 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+		LastViewedAt  int64
+	}
+
+	err = s.GetReplica().Select(&channels, queryString, args...)
```
Fixed in [`ed183f972d`](https://github.com/mattermost/mattermost/commit/ed183f972d).

agarciamontoro commented on server/channels/api4/channel_test.go:
```diff
@@ -4309,6 +4309,244 @@ func TestReadMultipleChannels(t *testing.T) {
+		_, _, err = client.CreatePost(context.Background(), &model.Post{
+			ChannelId: gmChannel.Id,
+			Message:   "test group message",
+		})
+		require.NoError(t, err)
+
+		channelResponse, _, err := client.ReadAllMessages(context.Background(), user.Id)
+		require.NoError(t, err)
+		require.Equal(t, "OK", channelResponse.Status, "invalid status return")
+		require.NotEmpty(t, channelResponse.LastViewedAtTimes, "should have viewed at times")
+	})
+
+	t.Run("Should fail marking messages for other user without permission", func(t *testing.T) {
+		_, _, err := client.ReadAllMessages(context.Background(), th.BasicUser2.Id)
+		require.Error(t, err)
+	})
+
+	t.Run("Admin should succeed in marking messages for other user", func(t *testing.T) {
+		adminClient := th.SystemAdminClient
+
+		dmChannel, _, err := adminClient.CreateDirectChannel(context.Background(), th.BasicUser2.Id, th.TeamAdminUser.Id)
+		require.NoError(t, err)
+
+		_, _, err = adminClient.CreatePost(context.Background(), &model.Post{
+			ChannelId: dmChannel.Id,
+			Message:   "test message for user2",
+		})
+		require.NoError(t, err)
+
+		channelResponse, _, err := adminClient.ReadAllMessages(context.Background(), th.BasicUser2.Id)
+		require.NoError(t, err)
+		require.Equal(t, "OK", channelResponse.Status, "invalid status return")
+		require.NotEmpty(t, channelResponse.LastViewedAtTimes, "should have viewed at times")
+	})
+
+	t.Run("Should handle empty direct/group message list gracefully", func(t *testing.T) {
+		channelResponse, _, err := client.ReadAllMessages(context.Background(), user.Id)
+		require.NoError(t, err)
+		require.Equal(t, "OK", channelResponse.Status, "invalid status return")
+	})
```
Not changing. Tests pass and exercise the feature.

agarciamontoro commented on server/channels/api4/channel_test.go:
```diff
@@ -4309,6 +4309,244 @@ func TestReadMultipleChannels(t *testing.T) {
+		require.Equal(t, http.StatusNotImplemented, resp.StatusCode)
+	})
+
+	t.Run("Should successfully mark all channels and threads as read for self in team", func(t *testing.T) {
+		channel, _, err := client.GetChannel(context.Background(), th.BasicChannel.Id)
+		require.NoError(t, err)
+		channel2, _, err := client.GetChannel(context.Background(), th.BasicChannel2.Id)
+		require.NoError(t, err)
+
+		post, _, err := client.CreatePost(context.Background(), &model.Post{
+			ChannelId: channel.Id,
+			Message:   "test message in channel 1",
+		})
+		require.NoError(t, err)
+
+		_, _, err = client.CreatePost(context.Background(), &model.Post{
+			ChannelId: channel2.Id,
+			Message:   "test message in channel 2",
+		})
+		require.NoError(t, err)
+
+		channelResponse, _, err := client.ReadAllInTeam(context.Background(), user.Id, team.Id)
+		require.NoError(t, err)
+		require.Equal(t, "OK", channelResponse.Status, "invalid status return")
+		require.NotEmpty(t, channelResponse.LastViewedAtTimes, "should have viewed at times")
+		require.Contains(t, channelResponse.LastViewedAtTimes, channel.Id)
+		require.GreaterOrEqual(t, channelResponse.LastViewedAtTimes[channel.Id], post.CreateAt,
+			"channel last_viewed_at should be at or after the latest post in the channel")
+	})
```
Not adding. Thread coverage exists in the storetest matrix.

agarciamontoro commented on server/channels/api4/channel_test.go:
```diff
@@ -4309,6 +4309,244 @@ func TestReadMultipleChannels(t *testing.T) {
+		require.NotEmpty(t, channelResponse.LastViewedAtTimes, "should have viewed at times")
+	})
+
+	t.Run("Should handle empty channel list gracefully", func(t *testing.T) {
+		newTeam := th.CreateTeam(t)
+		th.LinkUserToTeam(t, user, newTeam)
+
+		channelResponse, _, err := client.ReadAllInTeam(context.Background(), user.Id, newTeam.Id)
+
+		require.NoError(t, err)
+		require.Equal(t, "OK", channelResponse.Status, "invalid status return")
+		require.NotEmpty(t, channelResponse.LastViewedAtTimes, "should have viewed at times")
+	})
```
Not changing. Cosmetic naming nit.

agarciamontoro commented on server/channels/api4/channel.go:
```diff
@@ -603,6 +605,42 @@ func createDirectChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId()
+	if c.Err != nil {
+		return
+	}
+
+	auditRec := c.MakeAuditRecord(model.AuditEventMarkMessagesRead, model.AuditStatusFail)
+	defer c.LogAuditRec(auditRec)
+	model.AddEventParameterToAuditRec(auditRec, "user_id", c.Params.UserId)
+
+	if !c.App.SessionHasPermissionToUser(*c.AppContext.Session(), c.Params.UserId) {
+		c.SetPermissionError(model.PermissionEditOtherUsers)
+		return
+	}
+
+	times, err := c.App.MarkAllDirectAndGroupMessagesViewed(c.AppContext, c.Params.UserId, c.AppContext.Session().Id, c.App.IsCRTEnabledForUser(c.AppContext, c.Params.UserId))
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	resp := &model.ChannelViewResponse{
+		Status:            "OK",
+		LastViewedAtTimes: times,
+	}
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		c.Logger.Warn("Error while writing response", mlog.Err(err))
+	}
```
Fixed in [`89e7a69f31`](https://github.com/mattermost/mattermost/commit/89e7a69f31).

agarciamontoro commented on server/channels/app/channel.go:
```diff
@@ -3304,7 +3304,117 @@ func (a *App) SearchChannelsUserNotIn(rctx request.CTX, teamID string, userID st
+			message.Add("timestamp", timestamp)
+			a.Publish(message)
+		}
+	}
```
Fixed in [`27f075cf59`](https://github.com/mattermost/mattermost/commit/27f075cf59).

agarciamontoro commented on server/channels/store/sqlstore/channel_store.go:
```diff
@@ -2147,6 +2147,147 @@ func (s SqlChannelStore) GetChannelsWithUnreadsAndWithMentions(_ request.CTX, ch
+	err = s.GetReplica().Select(&channels, queryString, args...)
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "failed to find team channels with unreads and mentions data")
+	}
```
Fixed in [`8b651f1213`](https://github.com/mattermost/mattermost/commit/8b651f1213).

agarciamontoro commented on server/public/model/client4.go:
```diff
@@ -3265,6 +3265,25 @@ func (c *Client4) ReadMultipleChannels(ctx context.Context, userId string, chann
 	return DecodeJSONFromResponse[*ChannelViewResponse](r)
 }
 
+// ReadMultipleChannels performs a view action on all direct and group messages for a user
```
Fixed in [`fb09c9065b`](https://github.com/mattermost/mattermost/commit/fb09c9065b).

agarciamontoro commented on webapp/channels/src/components/feature_toast/features/mark_all_as_read_toast.tsx:
```diff
@@ -0,0 +1,75 @@
+        Preferences.CATEGORY_NEW_FEATURES,
+        Preferences.HAS_SEEN_MARK_ALL_READ_FEATURE,
+    );
+    const [show, setShow] = useState(
+        enableMarkAllReadShortcut &&
+        !UserAgent.isMobile() &&
+        !userHasSeenMarkAllReadToast,
+    );
```
Not adding. Feature flags do not toggle mid-session.

agarciamontoro commented on webapp/channels/src/components/mark_all_as_read_modal.test.tsx:
```diff
@@ -0,0 +1,120 @@
+        expect(onConfirm).toHaveBeenCalledWith(true);
+    });
+
+    test('should reset checkbox state when cancel is clicked', async () => {
+        renderWithContext(<MarkAllAsReadModal {...baseProps}/>);
+
+        const checkbox = screen.getByRole('checkbox', {name: /Don't ask me again/});
+        await userEvent.click(checkbox);
+        expect(checkbox).toBeChecked();
+
+        const cancelButton = screen.getByRole('button', {name: 'Cancel'});
+        await userEvent.click(cancelButton);
+
+        // Re-render to check state after cancel
+        const {rerender} = renderWithContext(<MarkAllAsReadModal {...baseProps}/>);
+        rerender(<MarkAllAsReadModal {...baseProps}/>);
+
+        const checkboxAfterCancel = screen.getByRole('checkbox', {name: /Don't ask me again/});
+        expect(checkboxAfterCancel).not.toBeChecked();
+    });
```
Not changing. Test passes.

agarciamontoro commented on webapp/channels/src/stores/hooks.ts:
```diff
@@ -45,3 +45,10 @@ export function useGlobalState<TVal>(
+ * existing class components.They can't use hooks,
+ * but will still want to have the same format as the hook.
+ */
+export const createStoredKey = (name: string, suffixToUse?: string) => `${name}${suffixToUse}`;
```
Fixed in [`ebbb9fd089`](https://github.com/mattermost/mattermost/commit/ebbb9fd089).

agarciamontoro committed:
Fix import order in shortcut_key consumers

agarciamontoro committed:
Fix CI

lieut-data reviewed (approved):
Thanks, @agarciamontoro!

agarciamontoro committed:
Merge remote-tracking branch 'upstream/master' into feature/MM-2541

# Conflicts:
#	api/v4/source/users.yaml

jgheithcock commented:
Woot! :tada:
